### PR TITLE
Bug 1119561 - [Cost Control][Settings] Reset Confirmation warning displa...

### PR DESCRIPTION
**Pull request by gitmai**
...ys inaccurate usage value the first time it is selected.

### CODE REVIEW

**Comment by try-server-hook** [comment]
[gitmai Marina (gitmai)](https://github.com/gitmai) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=273a06ede9c6)

**Comment by try-server-hook** [comment]
[gitmai Marina (gitmai)](https://github.com/gitmai) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=288cc3169bac)

File: apps/costcontrol/js/costcontrol.js [hunk 1]
```diff
@@ -617,8 +617,12 @@ var CostControl = (function() {
       result.status = 'success';
       result.data = lastDataUsage;
 
-      costcontrol.lastDataResults = lastDataUsage;
-
+      // Front-end workaround for Bug 1083680: Noticeable difference between by
+      // application breakdown totals and the total displayed in chart and
+      // widget (Exists traffic that cannot be not allocated to an app)
+      if (!perApp) {
```
Comment (lodr) [inline apps/costcontrol/js/costcontrol.js]: I would preffer to have two caches, one for aggregated query and another one for _by app_ query. What do you think? Although the second cache won't be used, at least we don't miss information.

File: apps/costcontrol/js/costcontrol.js [hunk 2]
```diff
@@ -617,8 +617,12 @@ var CostControl = (function() {
       result.status = 'success';
       result.data = lastDataUsage;
 
-      costcontrol.lastDataResults = lastDataUsage;
-
+      // Front-end workaround for Bug 1083680: Noticeable difference between by
```
Comment (lodr) [inline apps/costcontrol/js/costcontrol.js]: If we move to the [second cache approach](https://github.com/mozilla-b2g/gaia/pull/27347/files#r22933780), we could simplify this comment to:
```
// Once bug 1083680 is solved, both caches should contain the same info so we could use only `costcontrol.lastDataResults`.
```

File: apps/costcontrol/test/unit/cost_control_test.js [hunk 3]
```diff
@@ -498,4 +498,60 @@ suite('Cost Control Service Hub Suite >', function() {
       });
     }
   );
+
+    test(
```
Comment (lodr) [inline apps/costcontrol/test/unit/cost_control_test.js]: With the second cache approach, the test should become simpler as you only need to check the proper cache is filled with earch query type and the other remain the same.

**Comment by try-server-hook** [comment]
[gitmai Marina (gitmai)](https://github.com/gitmai) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=78dd8e24c102)

File: apps/costcontrol/test/unit/cost_control_test.js [hunk 4]
```diff
@@ -498,4 +498,60 @@ suite('Cost Control Service Hub Suite >', function() {
...
+    'Each data cache is refreshed depending of the request type',
+    function(done) {
+      var expectedLastDataUsage = {
+        mobile: { total: 4800543137 },
+        wifi: { total: 112123944 }
+      };
+
+      var fakeSettings = { lastDataUsage: expectedLastDataUsage };
+
+      sinon.stub(SimManager, 'requestDataSimIcc', function (callback) {
```
Comment (lodr) [inline apps/costcontrol/test/unit/cost_control_test.js]: Use `sinon.stub(SimManager, 'requestDataSimIcc').yields({issIc: '12345'});`

File: apps/costcontrol/test/unit/cost_control_test.js [hunk 5]
```diff
@@ -498,4 +498,60 @@ suite('Cost Control Service Hub Suite >', function() {
       });
     }
   );
+
+  test(
+    'Each data cache is refreshed depending of the request type',
```
Comment (lodr) [inline apps/costcontrol/test/unit/cost_control_test.js]: Split the testcase into 2, one for each cached value.

File: apps/costcontrol/test/unit/cost_control_test.js [hunk 6]
```diff
@@ -498,4 +498,60 @@ suite('Cost Control Service Hub Suite >', function() {
...
+        (typeof callback === 'function') && callback({iccId:'12345'});
+      });
+
+      window.ConfigManager = new MockConfigManager({
+        fakeSettings: fakeSettings
+      });
+
+      CostControl.getInstance(function(service) {
+        service.request({type: 'datausage'}, function(result) {
+          assert.equal(Object.keys(result.data.mobile.apps).length, 0);
```
Comment (lodr) [inline apps/costcontrol/test/unit/cost_control_test.js]: I would preffer the `deepEqual` comparison for empty objects.

File: apps/costcontrol/js/costcontrol.js (hunk 2 above)
Comment (lodr) [inline apps/costcontrol/js/costcontrol.js]: remember: change the comment, please.

File: apps/costcontrol/js/costcontrol.js (hunk 1 above)
Comment (lodr) [inline apps/costcontrol/js/costcontrol.js]: nit: invert `if` and `else` branches and express the condition in positive.

File: apps/costcontrol/test/unit/cost_control_test.js [hunk 7]
```diff
@@ -498,4 +498,60 @@ suite('Cost Control Service Hub Suite >', function() {
...
+          assert.equal(service.lastDataResults.mobile.total,
```
Comment (lodr) [inline apps/costcontrol/test/unit/cost_control_test.js]: When comparing, maybe you could create a couple of objects with the parts you want to compare and `deepEqual` those objects. You could do this in a custom `assertSomething` function.

**Comment by try-server-hook** [comment]
[gitmai Marina (gitmai)](https://github.com/gitmai) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=8964b107f989)

File: apps/costcontrol/test/unit/cost_control_test.js [hunk 8]
```diff
@@ -498,4 +498,75 @@ suite('Cost Control Service Hub Suite >', function() {
       });
     }
   );
+
+  test(
+    'Global Data cache is refreshed when the resquest is not per App',
```
Comment (lodr) [inline apps/costcontrol/test/unit/cost_control_test.js]: nit: the global word seems to suggest a global scoped data cache. Instead, lets describe the test as `Querying data usage globally caches the result`.

File: apps/costcontrol/test/unit/cost_control_test.js [hunk 9]
```diff
@@ -498,4 +498,75 @@ suite('Cost Control Service Hub Suite >', function() {
...
+      var fakeSettings = { lastDataUsage: {}, lastDataUsagePerApp: {} };
+      this.sinon.stub(SimManager, 'requestDataSimIcc').yields({iccId: '12345'});
+      window.ConfigManager = new MockConfigManager({
+        fakeSettings: fakeSettings
+      });
+
+      CostControl.getInstance(function(service) {
+        service.request({type: 'datausage'}, function(result) {
+          done(function() {
+            assert.deepEqual(result.data.mobile.apps, {});
```
Comment (lodr) [inline apps/costcontrol/test/unit/cost_control_test.js]: Actually, it suffices with making 2 assertions, one to compare `result` with `CostControl.lastDataUsage`. This one would mean "the value is properly cached and can be retrieved without another call to `request()`". And the second one consist into see that `lastDataUsagePerApp` is not destroyed. For this, you can manually set `lastDataUsagePerApp` to a well known object and then see it it refers to the same, after the calling. What do you think?

File: apps/costcontrol/test/unit/cost_control_test.js [hunk 10]
```diff
@@ -498,4 +498,75 @@ suite('Cost Control Service Hub Suite >', function() {
...
+        lastDataUsage: expectedLastDataUsage
+      };
+      this.sinon.stub(SimManager, 'requestDataSimIcc').yields({iccId: '12345'});
+      window.ConfigManager = new MockConfigManager({
+        fakeSettings: fakeSettings
+      });
+
+      CostControl.getInstance(function(service) {
+        // Request per app
+        var manifests = [MockMozNetworkStats.APP_MANIFEST_1,
```
Comment (lodr) [inline apps/costcontrol/test/unit/cost_control_test.js]: Same here.

**Comment by try-server-hook** [comment]
[gitmai Marina (gitmai)](https://github.com/gitmai) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=bf3c06aa4543)

**Comment by try-server-hook** [comment]
[gitmai Marina (gitmai)](https://github.com/gitmai) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=6f6f9348cc1a)

File: apps/costcontrol/test/unit/cost_control_test.js [hunk 11]
```diff
@@ -498,4 +498,41 @@ suite('Cost Control Service Hub Suite >', function() {
...
+                         MockMozNetworkStats.APP_MANIFEST_2];
+        var lastDataUsage = {};
+        service.lastDataUsage = lastDataUsage;
+        service.request({type: 'datausage', apps: manifests}, function(result) {
+          done(function() {
+            assert.equal(Object.keys(result.data.mobile.apps).length, 2);
```
Comment (lodr) [inline apps/costcontrol/test/unit/cost_control_test.js]: nit: I think this is unimportant.

**Comment by try-server-hook** [comment]
[gitmai Marina (gitmai)](https://github.com/gitmai) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=49c0a42030ae)

**Pull request merged**